_posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd

```diff
@@ -0,0 +1,347 @@
+manually compute the forward pass and the loss, and we still manually
+update the weights. Due to the latter, there is something I need to
+explain. But I'll let you check out the new version first:
+
+```{r}
+library(torch)
+
+### generate training data -----------------------------------------------------
+
+# input dimensionality (number of input features)
+d_in <- 3
+# output dimensionality (number of predicted features)
+d_out <- 1
+# number of observations in training set
+n <- 100
+
+
+# create random data
+x <- torch_randn(n, d_in)
+y <- x[, 1, NULL] * 0.2 - x[, 2, NULL] * 1.3 - x[, 3, NULL] * 0.5 + torch_randn(n, 1)
+
+
+### initialize weights ---------------------------------------------------------
+
+# dimensionality of hidden layer
+d_hidden <- 32
+# weights connecting input to hidden layer
+w1 <- torch_randn(d_in, d_hidden, requires_grad = TRUE)
+# weights connecting hidden to output layer
+w2 <- torch_randn(d_hidden, d_out, requires_grad = TRUE)
+
+# hidden layer bias
+b1 <- torch_zeros(1, d_hidden, requires_grad = TRUE)
+# output layer bias
+b2 <- torch_zeros(1, d_out, requires_grad = TRUE)
+
+### network parameters ---------------------------------------------------------
+
+learning_rate <- 1e-4
+
+### training loop --------------------------------------------------------------
+
+for (t in 1:200) {
+  ### -------- Forward pass --------
+  
+  y_pred <- x$mm(w1)$add(b1)$clamp(min = 0)$mm(w2)$add(b2)
+  
+  ### -------- compute loss -------- 
+  loss <- (y_pred - y)$pow(2)$sum()
+  if (t %% 10 == 0)
+    cat("Epoch: ", t, "   Loss: ", loss$item(), "\n")
+  
+  ### -------- Backpropagation --------
+  
+  # compute gradient of loss w.r.t. all tensors with requires_grad = TRUE
+  loss$backward()
+  
+  ### -------- Update weights -------- 
+  
+  # Wrap in with_no_grad() because this is a part we DON'T 
+  # want to record for automatic gradient computation
+   with_no_grad({
+     w1 <- w1$sub_(learning_rate * w1$grad)
+     w2 <- w2$sub_(learning_rate * w2$grad)
+     b1 <- b1$sub_(learning_rate * b1$grad)
+     b2 <- b2$sub_(learning_rate * b2$grad)  
+     
+     # Zero gradients after every pass, as they'd accumulate otherwise
+     w1$grad$zero_()
+     w2$grad$zero_()
+     b1$grad$zero_()
+     b2$grad$zero_()  
+   })
+
+}
+```
+
+As explained above, after `some_tensor$backward()`, all tensors
+preceding it in the graph[^2] will have their `grad` fields populated.
+We make use of these fields to update the weights. But now that
+*autograd* is "on", whenever we execute an operation we *don't* want
+recorded for backprop, we need to explicitly exempt it: This is why we
+wrap the weight updates in a call to `with_no_grad()`.
+
+[^2]: All that have `requires_grad` set to `TRUE`, to be precise.
+
+While this is something you may file under "nice to know" -- after all,
+once we arrive at the last post in the series, this manual updating of
+weights will be gone --, the idiom of *zeroing gradients* is here to
```

Comment:
```suggestion
weights will be gone -- the idiom of *zeroing gradients* is here to
```

Reply:
TL;DR of [this thread](https://english.stackexchange.com/questions/131874/can-i-write-a-comma-followed-by-an-em-dash) (and [this one](https://ell.stackexchange.com/questions/24960/can-a-comma-follow-an-em-dash#:~:text=The%20em%2Ddash%20does%20not,an%20abrupt%20change%20of%20thought), too): since an em dash replaces a comma, you don't put them together.

Reply:
That's very interesting and good to know! I'm pretty sure that would be different in German (but never googled it, i have to admit)